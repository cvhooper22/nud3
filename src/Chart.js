import React, { Component, PropTypes } from 'react';
import shallowCompare from 'react-addons-shallow-compare';
import _ from 'lodash';
import * as d3 from 'd3';
import dataDenormalizer from './helpers/dataDenormalizer';
import {
  stringOrFunc,
  arrayOrFunc,
  stringOrNumber,
} from './propTypes/customPropTypes';
import {
  updateYScaleDomain,
  updateXScaleDomain,
} from './helpers/scaleHelpers';
import splitHTMLElementProps from './helpers/splitHTMLElementProps';

export default class Chart extends Component {

  static propTypes = {
    DEBUG: PropTypes.bool,
    children: PropTypes.node,
    className: PropTypes.string,
    data: PropTypes.array,
    height: stringOrNumber,
    id: PropTypes.string,
    paddingBottom: PropTypes.number,
    paddingLeft: PropTypes.number,
    paddingRight: PropTypes.number,
    paddingTop: PropTypes.number,
    resizeDebounce: PropTypes.number,
    valueKeys: PropTypes.array,
    width: stringOrNumber,
    xKey: PropTypes.string,
    xScale: stringOrFunc,
    xScaleDomain: arrayOrFunc,
    xmlns: PropTypes.string.isRequired,
    yScale: stringOrFunc,
    yScaleDomain: arrayOrFunc,
    yScaleMaximum: PropTypes.number,
    yScaleMinimum: PropTypes.number,
    rangeRound: PropTypes.bool,
  };

  static defaultProps = {
    paddingBottom: 0,
    paddingLeft: 0,
    paddingRight: 0,
    paddingTop: 0,
    resizeDebounce: 100,
    valueKeys: ['value'],
    xScale: 'scaleTime',
    xmlns: 'http://www.w3.org/2000/svg',
    yScale: 'scaleLinear',
    rangeRound: false,
  };

  constructor (props, ...args) {
    super(props, ...args);
    this.onWindowResize = _.debounce(this.onWindowResize, props.resizeDebounce);
    this.setupData(props);
    this.state = {};
  }

  componentDidMount () {
    this.resize(this.props);
    window.addEventListener('resize', this.onWindowResize);
  }

  componentWillReceiveProps (nextProps) {
    this.setupData(nextProps);
    this.resize(nextProps);
  }

  shouldComponentUpdate (...args) {
    return shallowCompare(this, ...args);
  }

  componentWillUnmount () {
    window.removeEventListener('resize', this.onWindowResize);
  }

  onWindowResize = () => {
    if (this.node) {
      this.resize(this.props);
    }
  }

  setupData (props) {
    const data = this.calculateData(props);
    this.xScale = this.setupScale(props.xScale);
    this.yScale = this.setupScale(props.yScale);
    this.updateScalesFromData(props, data);
    this.chartData = data || [];
  }

  setupScale (prop) {
    return _.isFunction(prop) ? prop() : d3[prop]();
  }

  render () {
    const [elementProps] = splitHTMLElementProps(this.props);
    const style = { ...(elementProps.style || {}) };
    if (this.props.height) {
      style.height = this.sizeToStyle(this.props.height);
    }
    if (this.props.width) {
      style.width = this.sizeToStyle(this.props.width);
    }
    delete elementProps.data;
    delete elementProps.height;
    delete elementProps.width;
    delete elementProps.style;
    return (
      <svg
        { ...elementProps }
        style={ style }
        ref={ n => this.node = n }
        viewBox={ `0 0 ${this.state.width} ${this.state.height}` }
      >
        { this.renderChildren() }
      </svg>
    );
  }

  renderChildren () {
    if (!this.node) {
      return undefined;
    }
    return React.Children.map(this.props.children, this.renderChild);
  }

  renderChild = (child, i = 0) => {
    if (typeof child.type === 'string') {
      return child;
    }
    return React.cloneElement(child, {
      key: i,
      originalData: this.props.data,
      chartData: this.chartData,
      xScale: this.xScale,
      yScale: this.yScale,
      xScaleDomain: this.props.xScaleDomain,
      yScaleDomain: this.props.yScaleDomain,
      classNamePrefix: this.props.className,
      paddingBottom: this.props.paddingBottom,
      paddingLeft: this.props.paddingLeft,
      paddingRight: this.props.paddingRight,
      paddingTop: this.props.paddingTop,
      valueKeys: this.props.valueKeys,
      ...this.state,
      ...child.props,
      children: child.props.children,
      DEBUG: this.props.DEBUG,
    });
  }

  calculateData (props) {
    return this.calculateDenormalizedData(props.valueKeys || [], props.xKey, props.data || []);
  }

  calculateDenormalizedData (valueKeys, xKey, data) {
    return valueKeys.map(valueKey => dataDenormalizer(data || [], xKey, valueKey));
  }

  updateScalesFromData (props, data) {
    if (props.xKey) {
      updateXScaleDomain(this.xScale, props, data);
    }
    updateYScaleDomain(this.yScale, props, data);
  }

  resize (props) {
    if (!this.node) {
      return; // its debounced so its possible that its unmounted.
    }
    const rect = this.node.getBoundingClientRect();
    const height = props.height || rect.height;
    const width = props.width || rect.width;
    const areaWidth = width - (props.paddingLeft + props.paddingRight);
    const areaHeight = height - (props.paddingTop + props.paddingBottom);
    const rangeMethod = props.rangeRound ? 'rangeRound' : 'range';
    this.xScale[rangeMethod]([0, areaWidth]);
    this.yScale[rangeMethod]([areaHeight, 0]);
    this.setState({
      height,
      width,
      areaHeight,
      areaWidth,
    });
    if (props.DEBUG) {
      /* eslint-disable no-console */
      console.debug('xScale range:', [0, areaWidth]);
      console.debug('yScale range:', [areaHeight, 0]);
      console.debug(`settingSize: height:${height}, width:${width}, areaHeight:${areaHeight}, areaWidth:${areaWidth}`);
    }
  }

  sizeToStyle (size) {
    if (_.isString(size) && _.endsWith(size, '%')) {
      return size;
    }
    return `${size}px`;
  }
}
