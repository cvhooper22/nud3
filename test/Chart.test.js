import React from 'react';
import renderer from 'react-test-renderer';
import Chart from '../src/Chart';


describe('Chart.js', () => {
  it('should take proper html items and render them out', () => {
    const Fake = (props) => {
      return <path mask={ props.chartData } />;
    };
    const tree = renderer.create(
      <Chart data={ [] } style={{ backgroundColor: 'red' }} className="my-class">
        <Fake />
      </Chart>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should convert width and height to styles', () => {
    const tree = renderer.create(
      <Chart data={ [] } style={{ backgroundColor: 'red' }} width={ 999 } height={ 555 } />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should convert width and height percentages when strings', () => {
    const tree = renderer.create(
      <Chart data={ [] } style={{ backgroundColor: 'red' }} width="100%" height="55%" />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
