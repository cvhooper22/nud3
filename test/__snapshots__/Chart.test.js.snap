exports[`Chart.js should convert width and height percentages when strings 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "red",
      "height": "55%",
      "width": "100%",
    }
  }
  viewBox="0 0 undefined undefined" />
`;

exports[`Chart.js should convert width and height to styles 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "red",
      "height": "555px",
      "width": "999px",
    }
  }
  viewBox="0 0 undefined undefined" />
`;

exports[`Chart.js should take proper html items and render them out 1`] = `
<svg
  className="my-class"
  style={
    Object {
      "backgroundColor": "red",
    }
  }
  viewBox="0 0 undefined undefined" />
`;
